### ISSUE TITLE
Uploading file to default region results in a vague exception

### ISSUE THREAD

**Issue by KhaledSharif** [minio/minio#5455]
As suggested by the reply [to my issue on the boto repo](https://github.com/boto/boto3/issues/1425), I'm copying over the issue here. The reply suggests that `minio` is not accurately mimicking the AWS S3 response when no region is provided in the `boto` resource constructor.

# Error description

Trying to upload a file to the wrong S3 region results in a vague exception regarding an "invalid literal". This happens when the region is not specified in the Boto resource constructor. Specifying the region solves this error. If possible, it would be helpful to the user if this exception was made clearer.

# Environment

I'm using Python 3.5, with the following `boto` versions. The server is running `minio` ([Github repo](https://github.com/minio/minio)).

```
> pip3.5 list | grep boto

boto (2.48.0)
boto3 (1.5.20)
botocore (1.8.34)
```

# Exception trace

```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/response.py", line 397, in _update_chunk_length
    self.chunk_left = int(line, 16)
ValueError: invalid literal for int() with base 16: b''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "uploader.py", line 54, in <module>
    s3.Bucket("my-bucket").put_object(Key=file_key, Body=open(f, 'rb'), ContentType='image/png', ACL="public-read")
  File "/usr/local/lib/python3.5/dist-packages/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(**params)
  File "/usr/local/lib/python3.5/dist-packages/botocore/client.py", line 317, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/client.py", line 602, in _make_api_call
    operation_model, request_dict)
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 143, in make_request
    return self._send_request(request_dict, operation_model)
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 172, in _send_request
    success_response, exception):
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 265, in _needs_retry
    caught_exception=caught_exception, request_dict=request_dict)
  File "/usr/local/lib/python3.5/dist-packages/botocore/hooks.py", line 227, in emit
    return self._emit(event_name, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/hooks.py", line 210, in _emit
    response = handler(**kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 183, in __call__
    if self._checker(attempts, response, caught_exception):
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 251, in __call__
    caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 269, in _should_retry
    return self._checker(attempt_number, response, caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 317, in __call__
    caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 223, in __call__
    attempt_number, caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 359, in _check_caught_exception
    raise caught_exception
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 213, in _get_response
    proxies=self.proxies, timeout=self.timeout)
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/sessions.py", line 605, in send
    r.content
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/models.py", line 750, in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/models.py", line 673, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/response.py", line 303, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/response.py", line 447, in read_chunked
    self._update_chunk_length()
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/response.py", line 401, in _update_chunk_length
    raise httplib.IncompleteRead(line)
http.client.IncompleteRead: IncompleteRead(0 bytes read)
```

## Your Environment
```
VERSION
2018-01-18T20:33:21Z
MEMORY
Used: 3.4 MB | Allocated: 1.1 GB | Used-Heap: 3.4 MB | Allocated-Heap: 10 MB
PLATFORM
Host: c0eee334ae9d | OS: linux | Arch: amd64
RUNTIME
Version: go1.9.1 | CPUs: 8
```

**Comment by harshavardhana** [comment]
This has nothing to do with Minio. The execution is raised by boto on client side. It is a client side validation issue.

**Comment by harshavardhana** [comment]
Can you enable server environment variable MINIO_HTTP_TRACE=output.log and attach the log here ?

**Comment by harshavardhana** [comment]
```py
import boto                                                                                                                                                                   
import boto.s3.connection                                                                                                                                                     
                                                                                                                                                                              
conn = boto.connect_s3(                                                                                                                                                       
    aws_access_key_id = 'minio',                                                                                                                                              
    aws_secret_access_key = 'minio123',                                                                                                                                       
    host = 'localhost',                                                                                                                                                       
    port = 9000,                                                                                                                                                              
    is_secure = False,  
    calling_format = boto.s3.connection.OrdinaryCallingFormat(),                                                                                                                                              
)                                                                                                                                                                            
```

Is the configuration which i have used works with Minio in boto2. Please also paste your sample program @KhaledSharif  as well.

**Comment by KhaledSharif** [comment]
> This has nothing to do with Minio. The execution is raised by boto on client side. It is a client side validation issue.

I thought so too. However the reply on my issue on the `boto` repository seems to suggest otherwise. The reply claims that `minio` is not replicating the correct AWS S3 response.

> Can you enable server environment variable MINIO_HTTP_TRACE=output.log and attach the log here ?

Certainly. I'll check with our sysadmin and reply here with the log.

> Please also paste your sample program @KhaledSharif as well.

```python
s3 = boto3.resource(
    "s3",
    aws_access_key_id="ACCESS_KEY_ID",
    aws_secret_access_key="SECRET_ACCESS_KEY",
    endpoint_url="ENDPOINT",
    config=Config(signature_version="s3v4"),
    region_name="us-east-1",  # removing this line causes the exception!
)
```

**Comment by harshavardhana** [comment]
> Certainly. I'll check with our sysadmin and reply here with the log.

Were you able to get any more info? @KhaledSharif

**Comment by harshavardhana** [comment]
> Were you able to get any more info? @KhaledSharif

Reaching out again were you able to gather more server side info?

**Comment by KhaledSharif** [comment]
> Reaching out again were you able to gather more server side info?

Sorry for the delay in replying. I talked to our sysadmin and we can't modify the production server to output an HTTP trace (need to keep the production server as is). Instead I tried to reproduce the error locally using the following command, but the error cannot be reproduced, even though the version of `minio` is the same.

```
docker run --name minio -e MINIO_HTTP_TRACE=output.log -p 9000:9000 minio/minio:edge server /data
```

I'll be closing this issue until I'm able to reproduce it locally somehow. Thank you for your time @harshavardhana.
